src/aapns/connection.py

```diff
@@ -0,0 +1,440 @@
+                            self.protocol.increment_flow_control_window(
+                                event.flow_controlled_length
+                            )
+                        if ch:
+                            ch.ev.append(event)
+                            if not ch.fut.done():
+                                ch.fut.set_result(None)
+
+                # Somewhat inefficient: wake up background writer just in case
+                # it could be that we've received something that h2 needs to acknowledge
+                self.should_write.set()
+
+                # FIXME notify users about possible change to `.blocked`
+                # FIXME selective:
+                # * h2.events.WindowUpdated
+                # * max_concurrent_streams change
+                # * [maybe] starting a stream
+                # * a stream getting closed (but not half-closed)
+                # * closing / closed change
+        except ConnectionError as e:
+            if not self.outcome:
+                self.outcome = str(e)
+        except Exception:
+            logger.exception("background read task died")
+        finally:
+            self.closing = self.closed = True
+            for sid, ch in self.channels.items():
+                if ch.fut and not ch.fut.done():
+                    ch.fut.set_exception(Closed(self.outcome))
+
+    async def post(self, req: "Request") -> "Response":
+        assert len(req.body) <= MAX_NOTIFICATION_PAYLOAD_SIZE
+
+        now = time()
+        if now > req.deadline:
+            raise Timeout("Request timed out")
+        if self.closing or self.closed:
+            raise Closed(self.outcome)
+        if self.blocked:
+            raise Blocked()
+
+        try:
+            sid = self.protocol.get_next_available_stream_id()
+        except h2.NoAvailableStreamIDError:
+            self.closing = True
+            if not self.outcome:
+                self.outcome = "Exhausted"
+            raise Closed(self.outcome)
+
+        assert sid not in self.channels
+
+        self.last_new_sid = sid
+
+        ch = self.channels[sid] = Channel(sid, None, [])
+        self.protocol.send_headers(
+            sid, req.header_with(self.host, self.port), end_stream=False
+        )
+        # FIXME don't we have to increment global flow control
+        # also on reception of stream X data?
+        self.protocol.increment_flow_control_window(MAX_RESPONSE_SIZE, stream_id=sid)
+        self.protocol.send_data(sid, req.body, end_stream=True)
+        self.should_write.set()
+
+        try:
+            while not self.closed:
+                ch.fut = Future()
+                with suppress(TimeoutError):
+                    await wait_for(ch.fut, req.deadline - now)
+                now = time()
+                if now > req.deadline:
+                    raise Timeout()
+                for event in ch.ev:
+                    if isinstance(event, h2.events.ResponseReceived):
```

Comment:
how is this even working? there's no `import h2.events` (or `import h2`) anywhere I can see. what black magic makes this not error? could you add a `import h2.events` for the sake of my sanity please?

Reply:
wow, somehow `import h2.connection` magically populates `.events` in `h2`.
meanwhile `import h2.config` does not.

Thanks for pointing this out, it was quite breaky!

Reply:
hopefully done